### ISSUE TITLE
network problem caused by absent reverse entry in snat_v4_external map

### ISSUE THREAD

**Issue by ghost** [cilium/cilium#29305]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

This is a follow-up issue of #27821

This problem still exists in v1.14.4.

Recently I found that the problem can be triggered when lots of connections try to connect some targets with different local ports, steps to reproduced the problem are:

1. Run a server outside the cluster, listen some ports at same time. Assume the host addr is 99.99.99.99

```go
package main

import (
	"fmt"
	"net"
)

func main() {
	for i := 12345; i < 12355; i++ {
		go Listen(i)
	}
	c := make(chan struct{})
	<-c
}

func Listen(port int) {
	s := fmt.Sprintf("0.0.0.0:%d", port)
	fmt.Printf("start port %d\n", port)
	conn, err := net.Listen("tcp", s)
	if err != nil {
		panic(err)
	}

	for {
		netC, err := conn.Accept()
		if err != nil {
			fmt.Println(err)
		} else {
			fmt.Print("r")
			netC.Close()
		}
	}
}
```

2. create a client, put in into image. This client accesses all the listened ports for 1 minutes and sleep.

```go
package main

import (
	"flag"
	"fmt"
	"math/rand"
	"net"
	"strconv"
	"sync/atomic"
	"time"
)

var counter int32 = 0

func main() {
	var max int
	var target string
	var r bool
	flag.IntVar(&max, "max", 1000, "max parallism")
	flag.StringVar(&target, "target", "99.99.99.99", "addr")
	flag.BoolVar(&r, "rand", true, "random dest")
	flag.Parse()
	fmt.Printf("start with %d to %s\n", max, target)
	t := time.Tick(time.Minute)

out:
	for {
		select {
		case <-t:
			break out
		default:
			c := atomic.LoadInt32(&counter)
			if int(c) > max {
				continue
			}
			x := target + ":12345"
			if r {
				i := rand.Intn(10) + 12345
				x = target + ":" + strconv.Itoa(i)
			}

			go accessEndpoint(x)
		}
	}
	fmt.Println("done")
	time.Sleep(time.Hour * 30)
}

func accessEndpoint(target string) {
	atomic.AddInt32(&counter, 1)
	d := net.Dialer{Timeout: time.Second * 10}
	conn, err := d.Dial("tcp", target)
	if err != nil {
		info := ""
		if conn != nil {
			info = conn.LocalAddr().String() //this info is useless 
		}
		fmt.Printf("%s [%s]\n", err, info)
	} else {
		conn.Close()
	}
	atomic.AddInt32(&counter, -1)
}
```

3. In our machine, the NAT map size is 131072, which can be set with `bpf-nat-global-max`

4. before creating the pod, flush nat map of one node, enter the agent and

```shell
cilium bpf nat flush
```

5. create the pod on that node

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: client
  namespace: <ns>
spec:
  containers:
  - image: <image>
    name: <name>
  nodeSelector:
    kubernetes.io/hostname: <node>
```

the server will continuously print `r`

6. after server stop printing, examine the nat map, check the entries that is OUT/egress, but no corresponding IN/ingress reversed nat entry. Get the source port in the OUT item and target addr and port, which can be done with `cilium bpf nat list`. I create an ebpf map manipulation tool to take care of cilium ebpf map, so I don't put the command here. The steps are like:

```
 - dump all item |grep addr | grep OUT | awk the nat addr and port > out.txt
 - dump all item |grep addr | grep IN > in.txt
 - for i in `cat out.txt`; do res=`cat in.txt |grep addr |grep $i`; if [ "$res" == "" ]; then echo $i; fi; done
  several items are printed, choose one and then
 - dump all item |grep the choose one |grep OUT
```

7. enter the pod, access the target with the local port from 6

```
nc -vzp <source port> <target addr> <target port>
```

the command stucks. However If we execute the same command on the node of host pod or if we change another local port, it succeeds.

8. check the prometheus metrics `cilium_datapath_signals_handled_total`, get nothing

```shell
curl 127.0.0.1:9090/metrics | grep nat_fill_up
```

Recently I noticed that the comment of #28857 says

```go
// Typically NAT entries should get removed along with their owning CT entry,
// as part of purgeCtEntry*(). But stale NAT entries can get left behind if the
// CT entry disappears for other reasons - for instance by LRU eviction, or
// when the datapath re-purposes the CT entry.
```

And in this [link](https://stackoverflow.com/questions/75882443/elements-incorrectly-evicted-from-ebpf-lru-hash-map), someone proved that LRU map can evict item even far from full. And the removed way is not strictly LRU.

I try to change the `NAT_MAP_TYPE` to `BPF_MAP_TYPE_HASH` and retry the test, although the `snat_nat_fill` is sent tens of times, the problem is solved.

In the `PurgeOrphanNATEntries()` of v1.11.17, only `TUPLE_F_IN` entries is checked, if `TUPLE_F_IN` is not exists, the `TUPLE_F_OUT` item will not be cleared at the time. The `PurgeOrphanNATEntries()` of v1.14.4 fix this. However if no signal `snat_nat_fill` is sent, `PurgeOrphanNATEntries()` will not be invoked, so the problem will last for a period.

### Cilium Version

v1.14.4 (and v1.11.17)

### Kernel Version

5.10

### Kubernetes Version

1.24.3

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by brb** [comment]
Thanks for the very detailed report!

>However if no signal snat_nat_fill is sent, PurgeOrphanNATEntries() will not be invoked, so the problem will last for a period.

Could we run `PurgeOrphanNATEntries()` periodically?

**Comment by nijich** [comment]
Hi, I think it is not a perfect solution. There is still a time gap between the eviction and periodically clean.

**Comment by nijich** [comment]
Another thing I want to mention is, if I disable the bpf-masquerade, the problem will take place on node. This is because even disable the option, the NAT map still keep `HOST_LOCAL` entries, which is a little confusing for me.

**Comment by nijich** [comment]
More information here: I observed the problem mentioned in #27405 happened before the network problem happened in one of my cases. 

It looks like:
1. several `nat_fill_up` signals sent, and causing GC Interval's calculation wrong and up to 12 hour
2. expired ct and nat entries left not to be cleared
3. entries become more and more, fill the space of nat map, start to evicted entries due to LRU
4. the otuple entry is not evicted, but is rtuple is evicted.

Besides, the conntrack entry may also be evicted by LRU, which cause the otuple entries left until LRU decides to clear it. However, if the application try to use the same port to the same destination, the otuple entries will be hit again and may not be evicted.

**Comment by nijich** [comment]
#27405 also causes my other issue: #28455

**Comment by liuyuan10** [comment]
> And in this [link](https://stackoverflow.com/questions/75882443/elements-incorrectly-evicted-from-ebpf-lru-hash-map), someone proved that LRU map can evict item even far from full. And the removed way is not strictly LRU.

Isn't this the key problem here instead of PurgeOrphanNATEntries? With LRU randomly evict entires, it may remove IN or OUT entries without removing the paired one leaving single entry breaking traffic. 

> I try to change the NAT_MAP_TYPE to BPF_MAP_TYPE_HASH and retry the test, although the snat_nat_fill is sent tens of times, the problem is solved.

This exactly proves that the LRU behavior is the key here but not PurgeOrphanNATEntries

maybe the fix is that whenever an OUT entry is hit, we check and create the IN entry if it's missing

**Comment by nijich** [comment]
Yes, the intrinsic problem is the LRU eviction. And other problems just contribute to connection failure.

**Comment by sugangli** [comment]
If we look at the __snat_update call:
https://github.com/cilium/cilium/blob/76a659c64b58ec3cff7b91c8eb16de2a3fa5ef42/bpf/lib/nat.h#L69-L74

The existing entry won't be updated as we pass in `BPF_NOEXIST`.  So if a `otuple` is corrupted, it's `rtuple` won't be updated, and this `otuple` won't be recovered.  This becomes a problem when a connection is pinned to a IP:port.

**Comment by julianwiedmann** [comment]
Drive-by comment - it's really hard to follow this issue and/or recommend next steps. I'd suggest to open **separate** issues for every aspect you encounter, instead of mixing everything into a single issue.

Just browsing through I see
* mention of a GC bug that was presumably fixed by https://github.com/cilium/cilium/pull/28657
* mention of issues on `v1.11` (which is long EOL)
* recovery from LRU eviction of NAT entries (do we need separate discussion for eviction of the EGRESS and INGRESS entry?)
* failure to allocate ports after `SNAT_COLLISION_RETRIES`
* some misbehaviour of `__snat_update()` ?

**Comment by nijich** [comment]
Hi, your summary is precise. I think we can only care about the LRU problem here. I noticed the `SNAT_COLLISION_RETRIES` problem is already counted in #31643. The `GC interval` and `PurgeOrphan()` bugs are already fixed now, which is only used to provide possibly related background. No need to care about `v1.11`.

>
>     * recovery from LRU eviction of NAT entries (do we need separate discussion for eviction of the EGRESS and INGRESS entry?)
>
In my limited usage situations, absent `TUPLE_F_OUT` entry will not affect the network function. I don't know if it can affect some features like DSR.

>
>     * some misbehaviour of `__snat_update()` ?
>
I think the comment from @sugangli  about modify `__snat_update()` is for solving the absent entry problem. I think maybe the modification is useful but I don't know if it works properly.

**Comment by sugangli** [comment]
```
* recovery from LRU eviction of NAT entries (do we need separate discussion for eviction of the EGRESS and INGRESS entry?)
```

We also have a reproduction for the ingress case which show the same symptom.  The recovery mechanism should be similar. Maybe we could keep them in the same discussion?

```
* some misbehaviour of `__snat_update()` ?
``` 

That's just the snippet which I think why the evicted entry cannot be recovered.

**Comment by sugangli** [comment]
In our experiment, where we have two clusters (client and server), and we deploy a ILB in server cluster. 

We then generated lots of curl traffic from the client  cluster (20 pods) to the server with 1 backend through the ILB.  
Then we dumped the ct maps and the nat maps on the server sides every 1 minutes. At the same time, we checked the nat maps to see if a `TCP OUT` is missing its `TCP IN`.  If there was, we recorded the tuples.  Then we searched in the CT maps, and was able to find the corresponding CT entries that were still alive.

For example, a broken NAT entry that misses it TCP IN:
```
TCP OUT 10.88.5.88:51825 -> 10.80.2.5:80 XLATE_SRC 10.128.0.14:59344 Created=142605sec NeedsCT=0
```
Its CT entry:
```
TCP OUT 10.88.5.88:51825 -> 10.80.2.5:80 expires=17333057 RxPackets=5 RxBytes=1191 RxFlagsSeen=0x1b LastRxReport=17333018 TxPackets=7 TxBytes=546 TxFlagsSeen=0x1b LastTxReport=17333017 Flags=0x0033 [ RxClosing TxClosing SeenNonSyn NodePort ] RevNAT=13 SourceSecurityID=2 IfIndex=2
```

In addition, we were able to observe the misrouting issue due to this once in a while:

```
# Client pod to ILB SYN
16:33:53.681878 IP 10.88.4.85.49183 > 10.128.0.47.http: Flags [S], seq 3440461061, win 42600, options [mss 1420,sackOK,TS val 1322117870 ecr 0,nop,wscale 7], length 0

# LB node to server pod SYN
16:33:53.681928 IP gke-psc-server-default-pool-63802cd2-959j.us-central1-c.c.sugangli-gke-dev.internal.49183 > 10.80.0.13.http: Flags [S], seq 3440461061, win 42600, options [mss 1420,sackOK,TS val 1322117870 ecr 0,nop,wscale 7], length 0

# server pod to LB Node SYN|ACK
16:33:53.682144 IP 10.80.0.13.http > gke-psc-server-default-pool-63802cd2-959j.us-central1-c.c.sugangli-gke-dev.internal.49183: Flags [S.], seq 1895312037, ack 3440461062, win 43648, options [mss 1420,sackOK,TS val 1037016690 ecr 1322117870,nop,wscale 7], length 0

# ILB to client pod SYN|ACK is MISSING !!!!!

```

I drafted a WIP PR that fixed this path. Let me know if it makes sense in general.
